contracts/connection-router/src/contract/execute.rs

```diff
@@ -126,46 +146,461 @@ where
             return Err(report!(ContractError::WrongSourceChain));
         }
 
-        let mut wasm_msgs = vec![];
-
-        for (destination_chain, msgs) in &msgs.iter().group_by(|msg| msg.destination_chain.clone())
-        {
-            let destination_chain = self
-                .store
-                .load_chain_by_chain_name(&destination_chain)?
-                .ok_or(ContractError::ChainNotFound)?;
-            if destination_chain.outgoing_frozen() {
-                return Err(report!(ContractError::ChainFrozen {
-                    chain: destination_chain.name,
-                }));
-            }
-
-            wasm_msgs.push(WasmMsg::Execute {
-                contract_addr: destination_chain.gateway.address.to_string(),
-                msg: to_binary(&ExecuteMsg::RouteMessages(msgs.cloned().collect()))
-                    .expect("must serialize message"),
-                funds: vec![],
-            });
-        }
+        Ok(())
+    }
+
+    pub fn route_messages(
+        self,
+        sender: Addr,
+        msgs: Vec<Message>,
+    ) -> error_stack::Result<Response, ContractError> {
+        self.validate_msgs(&sender, &msgs)?;
+
+        let wasm_msgs = msgs
+            .iter()
+            .group_by(|msg| msg.destination_chain.to_owned())
+            .into_iter()
+            .map(|(destination_chain, msgs)| {
+                let gateway = match self.store.load_chain_by_chain_name(&destination_chain)? {
+                    Some(destination_chain) => {
+                        if destination_chain.outgoing_frozen() {
+                            return Err(report!(ContractError::ChainFrozen {
+                                chain: destination_chain.name,
+                            }));
+                        }
+
+                        destination_chain.gateway.address
+                    }
```

Comment:
make these two separate cases by moving the if clause into the pattern match?

Reply:
Like that